### PR TITLE
Update tests after Account refactor

### PR DESCRIPTION
**Pull request by cptartur**
### **Please check if the PR fulfills these requirements**
- [ ] The formatter, linter, and tests all run without an error
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

## **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

**This PR doesn't include a update for docs tests, this will be a separate PR**

* **What is the current behavior?** (You can also link to an open issue here)

### CODE REVIEW

File: starknet_py/tests/e2e/account/account_client_test.py [hunk 1]
```diff
@@ -74,12 +74,14 @@ async def test_get_balance_default_token_address(net):
...
     with patch(
-        "starknet_py.net.account.account_client.AccountClient.estimate_fee", MagicMock()
+        # "starknet_py.net.account.account_client.AccountClient.estimate_fee", MagicMock()
+        "starknet_py.net.gateway_client.GatewayClient.estimate_fee",
+        AsyncMock(),
     ) as mocked_estimate_fee:
-        result = asyncio.Future()
-        result.set_result([0])
+        # result = asyncio.Future()
+        # result.set_result([0])
```
Comment (war-in) [inline starknet_py/tests/e2e/account/account_client_test.py:82]: 👀

File: starknet_py/tests/e2e/account/account_client_test.py [hunk 2]
```diff
@@ -186,14 +188,14 @@ async def test_get_class_hash_at(map_contract, account_client):
...
+# async def test_throws_on_wrong_transaction_version(new_deploy_map_contract):
+#     with pytest.raises(ValueError) as err:
+#         await new_deploy_map_contract.functions["put"].invoke(
+#             key=10, value=20, version=0, max_fee=MAX_FEE
+#         )
+#
+#     assert (
+#         "Provided version: 0 is not equal to account's supported_tx_version: 1"
+#         in str(err.value)
+#     )
```
Comment (war-in) [inline starknet_py/tests/e2e/account/account_client_test.py:201]: 👀

File: starknet_py/tests/e2e/contract_interaction/interaction_test.py [hunk 3]
```diff
@@ -166,56 +165,51 @@ async def test_call_uninitialized_contract(gateway_account_client):
 
 
 @pytest.mark.asyncio
-async def test_wait_for_tx(account_client, map_contract):
+async def test_wait_for_tx(gateway_client, map_contract):
```
Comment (war-in) [inline starknet_py/tests/e2e/contract_interaction/interaction_test.py:168]: why do we use `gateway_client` here?

Comment (war-in) [reply starknet_py/tests/e2e/contract_interaction/interaction_test.py:168]: Let's use the `client`

File: starknet_py/tests/e2e/contract_interaction/interaction_test.py [hunk 4]
```diff
@@ -166,56 +165,51 @@ async def test_call_uninitialized_contract(gateway_account_client):
...
+# async def test_warning_when_max_fee_equals_to_zero(map_contract):
+#     with pytest.warns(
+#         DeprecationWarning,
+#         match=r"Transaction will fail with max_fee set to 0. Change it to a higher value.",
+#     ):
+#         # try except have to be added because when running on a real environment it will throw an error (max_fee=0)
+#         try:
+#             await map_contract.functions["put"].invoke(10, 20, max_fee=0)
+#         except ClientError:
+#             pass
```
Comment (war-in) [inline starknet_py/tests/e2e/contract_interaction/interaction_test.py:200]: 👀

File: starknet_py/tests/e2e/fixtures/account_clients.py [hunk 5]
```diff
@@ -292,6 +294,70 @@ def new_account_client(request) -> AccountClient:
     return request.getfixturevalue(request.param)
 
 
+@pytest.fixture(scope="module")
+def gateway_account(
+    new_address_and_private_key: Tuple[str, str], gateway_client: GatewayClient
+) -> BaseAccount:
+    """
+    Returns a new AccountClient created with FullNodeClient
```
Comment (war-in) [inline starknet_py/tests/e2e/fixtures/account_clients.py:302]:
```suggestion
    Returns a new Account created with GatewayClient
```

File: starknet_py/tests/e2e/fixtures/account_clients.py [hunk 6]
```diff
@@ -292,6 +294,70 @@ def new_account_client(request) -> AccountClient:
...
+def full_node_account_proxy(new_full_node_account_client: AccountClient) -> BaseAccount:
+    return AccountProxy(new_full_node_account_client)
+
+
+@pytest.fixture(scope="module")
+def full_node_account(
+    new_address_and_private_key: Tuple[str, str], full_node_client: FullNodeClient
+) -> BaseAccount:
+    """
+    Returns a new AccountClient created with FullNodeClient
```
Comment (war-in) [inline starknet_py/tests/e2e/fixtures/account_clients.py:329]:
```suggestion
    Returns a new Account created with FullNodeClient
```

File: starknet_py/tests/e2e/fixtures/account_clients.py [hunk 7]
```diff
@@ -292,6 +294,70 @@ def new_account_client(request) -> AccountClient:
...
+    return accounts
+
+
+@pytest.fixture(
+    scope="module",
+    params=net_to_base_accounts(),
+)
+def account(request) -> BaseAccount:
+    """
+    This parametrized fixture returns all new AccountClients, one by one.
```
Comment (war-in) [inline starknet_py/tests/e2e/fixtures/account_clients.py:356]:
```suggestion
    This parametrized fixture returns all new Accounts, one by one.
```

**Review by war-in — CHANGES_REQUESTED** [review]

File: starknet_py/net/account/account_test.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+from starknet_py.contract import Contract
+from starknet_py.net.client_models import Call
+from starknet_py.tests.e2e.utils import MAX_FEE
+
+
+@pytest.mark.asyncio()
+async def test_get_nonce(gateway_account):
+    nonce = await gateway_account.get_nonce()
+
+    assert nonce >= 0  # TODO maybe better test
```
Comment (war-in) [inline starknet_py/net/account/account_test.py:12]: What about e2e tests here?

File: starknet_py/tests/e2e/account/account_test.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    ) as mocked_call_contract:
+        mocked_call_contract.return_value = [0, 0]
+
+        await acc_client.get_balance()
+
+        call = mocked_call_contract.call_args
+
+    (call,) = call[0]
+
+    assert call.to_addr == parse_address(FEE_CONTRACT_ADDRESS)
```
Comment (Solpatium) [inline starknet_py/tests/e2e/account/account_test.py:68]: This is a unit test, isn't it?

**Review by Solpatium — APPROVED** [review]
LGTM with comment 👌

File: starknet_py/tests/e2e/account/account_test.py (hunk 9 above)
Comment (cptartur) [reply starknet_py/tests/e2e/account/account_test.py:68]: It's not using a real network, so I think you are right.

**Comment by cptartur** [comment]
What still needs addressing: my ide is complaining about `typed_data` not being of correct type, even though tests and typecheck pass: This is probably caused by the changes in Signer interface.

File: starknet_py/net/account/account_test.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+from starknet_py.net.models import StarknetChainId, parse_address
+from starknet_py.net.networks import TESTNET, MAINNET
+from starknet_py.net.signer.stark_curve_signer import StarkCurveSigner
+
+
+def test_contract_raises_on_no_client_and_account():
+    with pytest.raises(ValueError) as exinfo:
+        Contract(address=1234, abi=[])
+
+    assert "One of client or account must be provided" in str(exinfo.value)
```
Comment (drknzz) [inline starknet_py/net/account/account_test.py:19]: We could maybe start using the `match` keyword argument
```
def test_match():
    with pytest.raises(ValueError, match=r".* 123 .*"):
        myfunc()
```

File: starknet_py/tests/e2e/fixtures/account_clients.py [hunk 11]
```diff
@@ -292,6 +294,70 @@ def new_account_client(request) -> AccountClient:
...
+) -> BaseAccount:
+    """
+    Returns a new Account created with GatewayClient
+    """
+    address, private_key = new_address_and_private_key
+
+    return Account(
+        address=address,
+        client=gateway_client,
+        key_pair=KeyPair.from_private_key(int(private_key, 0)),
```
Comment (drknzz) [inline starknet_py/tests/e2e/fixtures/account_clients.py:309]: Cool 👍 didn't know about 0 being a possible base.

File: starknet_py/tests/e2e/fixtures/account_clients.py [hunk 12]
```diff
@@ -292,6 +294,70 @@ def new_account_client(request) -> AccountClient:
     return request.getfixturevalue(request.param)
 
 
+@pytest.fixture(scope="module")
+def gateway_account(
+    new_address_and_private_key: Tuple[str, str], gateway_client: GatewayClient
+) -> BaseAccount:
+    """
+    Returns a new Account created with GatewayClient
```
Comment (drknzz) [inline starknet_py/tests/e2e/fixtures/account_clients.py:302]: Are we commited to putting dots after every sentence in docstrings?

**Review by drknzz — COMMENTED** [review]
[This](https://github.com/software-mansion/starknet.py/pull/551/files?diff=unified&w=0#file-starknet_py-net-signer-stark_curve_signer-py-L125) sounds like the whole TypedData class as TypedDict has been deprecated. It's not the case, is it?

Maybe we could change it to `"typed_data as dict has been deprecated"`.

**Comment by drknzz** [comment]
> What still needs addressing: my ide is complaining about `typed_data` not being of correct type, even though tests and typecheck pass: This is probably caused by the changes in Signer interface.

I believe changing
```
from starknet_py.utils.typed_data import TypedData
```
to
```
from starknet_py.net.models.typed_data import TypedData
```

fixes the issue.

File: starknet_py/tests/e2e/fixtures/account_clients.py (hunk 12 above)
Comment (cptartur) [reply starknet_py/tests/e2e/fixtures/account_clients.py:302]: We should be, but I forget at times. Will fix that.

**Comment by cptartur** [comment]
> [This](https://github.com/software-mansion/starknet.py/pull/551/files?diff=unified&w=0#file-starknet_py-net-signer-stark_curve_signer-py-L125) sounds like the whole TypedData class as TypedDict has been deprecated. It's not the case, is it?
> 
> Maybe we could change it to `"typed_data as dict has been deprecated"`.

I think we can change it, it does sound a bit more descriptive.



> > What still needs addressing: my ide is complaining about `typed_data` not being of correct type, even though tests and typecheck pass: This is probably caused by the changes in Signer interface.
> 
> I believe changing
> 
> ```
> from starknet_py.utils.typed_data import TypedData
> ```
> 
> to
> 
> ```
> from starknet_py.net.models.typed_data import TypedData
> ```
> 
> fixes the issue.

I'll check that 👍

File: starknet_py/tests/e2e/account/account_client_test.py [hunk 13]
```diff
@@ -187,13 +186,25 @@ async def test_get_class_hash_at(map_contract, account_client):
...
 
     assert (
         "Provided version: 0 is not equal to account's supported_tx_version: 1"
-        in str(err.value)
+        in str(exinfo.value)
     )
+
+
+@pytest.mark.asyncio
+async def test_sign_wanrs_on_max_fee_0(account_client):
```
Comment (war-in) [inline starknet_py/tests/e2e/account/account_client_test.py:202]:
```suggestion
async def test_sign_warns_on_max_fee_0(account_client):
```

File: starknet_py/net/account/account_test.py [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+
+        call = mocked_call_contract.call_args
+
+    (call,) = call[0]
+
+    assert call.to_addr == parse_address(FEE_CONTRACT_ADDRESS)
+
+
+@pytest.mark.asyncio
+async def test_create_account():
```
Comment (war-in) [inline starknet_py/net/account/account_test.py:49]: We don't need `async` here

File: starknet_py/net/account/account_test.py [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+        key_pair=key_pair,
+        chain=StarknetChainId.TESTNET,
+    )
+
+    assert account.address == 0x1
+    assert account.signer.public_key == key_pair.public_key
+
+
+@pytest.mark.asyncio
+async def test_create_account_from_signer():
```
Comment (war-in) [inline starknet_py/net/account/account_test.py:63]: and here

File: starknet_py/net/account/account_test.py [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+        chain_id=StarknetChainId.TESTNET,
+    )
+    account = Account(address=0x1, client=GatewayClient(net=TESTNET), signer=signer)
+
+    assert account.address == 0x1
+    assert account.signer == signer
+
+
+@pytest.mark.asyncio
+async def test_create_account_raises_on_no_chain_and_signer():
```
Comment (war-in) [inline starknet_py/net/account/account_test.py:76]: and here

File: starknet_py/net/account/account_test.py [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+            address=0x1,
+            client=GatewayClient(net=TESTNET),
+            key_pair=KeyPair.from_private_key(0x111),
+        )
+
+    assert "One of chain or signer must be provided" in str(exinfo.value)
+
+
+@pytest.mark.asyncio
+async def test_create_account_raises_on_no_keypair_and_signer():
```
Comment (war-in) [inline starknet_py/net/account/account_test.py:86]: here too

File: starknet_py/net/account/account_test.py [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+        )
+
+    assert (
+        "Either a signer or a key_pair must be provided in AccountClient constructor"
+        in str(exinfo.value)
+    )
+
+
+@pytest.mark.asyncio
+async def test_create_account_raises_on_both_keypair_and_signer():
```
Comment (war-in) [inline starknet_py/net/account/account_test.py:99]: 🌕 👋

File: starknet_py/tests/e2e/account/account_test.py [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+    )
+
+    assert class_hash != 0
+
+
+@pytest.mark.asyncio()
+async def test_get_nonce(gateway_account):
+    nonce = await gateway_account.get_nonce()
+
+    assert nonce >= 0  # TODO maybe better test
```
Comment (war-in) [inline starknet_py/tests/e2e/account/account_test.py:139]: 👀

File: starknet_py/tests/e2e/account/account_test.py [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+
+from starknet_py.contract import Contract
+from starknet_py.net.client_models import Call
+from starknet_py.net.gateway_client import GatewayClient
+from starknet_py.tests.e2e.fixtures.constants import MAX_FEE
+from starknet_py.transaction_exceptions import TransactionRejectedError
+
+
+@pytest.mark.asyncio
+async def test_deploy_account_contract_and_sign_tx(map_contract):
```
[the remark this answers is not in the document]
Comment (cptartur) [reply starknet_py/tests/e2e/account/account_test.py:11]: Let's remove it entirely: It's duplicating contract tests we already have.

File: starknet_py/tests/e2e/account/account_test.py (hunk 19 above)
Comment (cptartur) [reply starknet_py/tests/e2e/account/account_test.py:128]: I wonder if I should change this test?

File: starknet_py/tests/e2e/account/account_client_test.py [hunk 21]
```diff
@@ -187,13 +175,25 @@ async def test_get_class_hash_at(map_contract, account_client):
 
 
 @pytest.mark.asyncio
-async def test_throws_on_wrong_transaction_version(new_deploy_map_contract):
-    with pytest.raises(ValueError) as err:
-        await new_deploy_map_contract.functions["put"].invoke(
-            key=10, value=20, version=0, max_fee=MAX_FEE
+async def test_sign_transaction_unsupported_version(new_account_client):
+    with pytest.raises(ValueError) as exinfo:
```
Comment (drknzz) [inline starknet_py/tests/e2e/account/account_client_test.py:179]: Could we use match?

File: starknet_py/tests/e2e/contract_interaction/interaction_test.py [hunk 22]
```diff
@@ -273,17 +252,22 @@ async def test_general_simplified_deployment_flow(
...
     await deploy_result.wait_for_acceptance()
 
     contract = deploy_result.deployed_contract
 
     assert isinstance(contract.address, int)
     assert len(contract.functions) != 0
+
+
+def test_contract_raises_on_both_client_and_account(gateway_client, gateway_account):
+    with pytest.raises(ValueError) as exinfo:
```
Comment (drknzz) [inline starknet_py/tests/e2e/contract_interaction/interaction_test.py:270]: Same here

**Review by drknzz — APPROVED** [review]
🥇

File: starknet_py/tests/e2e/account/account_test.py (hunk 19 above)
Comment (drknzz) [reply starknet_py/tests/e2e/account/account_test.py:128]: I don't have any idea to test it any better while staying simple, so I'd say it's ok as it is.

Comment (war-in) [reply starknet_py/tests/e2e/account/account_test.py:128]: We do 😎 @cptartur will add it soon

**Review by war-in — APPROVED** [review]

**Pull request merged**